pkg/cmd/recycle/walker.go

```diff
@@ -0,0 +1,135 @@
+package recycle
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"sort"
+)
+
+// Visits a directory tree, depth-first, calling walkFn for every file/directory
+type Walker struct {
+	walkFn filepath.WalkFunc
+
+	// fsuid holds the last fsuid we set
+	fsuid int64
+
+	// lstat is for testing, defaults to os.Lstat
+	lstat func(name string) (fi os.FileInfo, err error)
+
+	// setfsuid is for testing, defaults to syscall.Setfsuid
+	setfsuid func(uid int) (err error)
+
+	// getuid is for testing, defaults to fileinfo.Sys().(*syscall.Stat_t).Uid
+	getuid func(os.FileInfo) (int64, error)
+}
+
+func NewWalker(walkFn filepath.WalkFunc) *Walker {
+	return &Walker{
+		walkFn:   walkFn,
+		fsuid:    -1,
+		lstat:    os.Lstat,
+		setfsuid: setfsuid,
+		getuid:   getuid,
+	}
+}
+
+func (w *Walker) Walk(root string) error {
+	// Lock threads, so our Setfsuid calls always apply to the same thread
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// TODO: determine original fsuid and switch back?
+
+	// The launching process must have stat permission on the root dir
+	info, err := w.lstat(root)
+	if err != nil {
+		return w.walkFn(root, info, err)
+	}
+	return w.walk(root, info)
```

Comment:
> What happens if there are a million files? Is there some limit where we'd need to pull out before we're crippling the node or IO or something?

This runs in a pod.  Seems like we'd want to solve that the same way we'd solve it for a pod.